### PR TITLE
Bugfix/litellm plus generic exceptions

### PR DESCRIPTION
**Pull request by bhancockio**
Issue:
- We updated our agent and crew_agent_executor to track when LiteLLM threw an error. Originally, I thought there was only one type of error - BaseLLMException. However, different models throw different errors.
- Because we weren't properly catching errors from Anthropic and other models, our invoke_loop logic ends up in a loop where it's making many LLM calls that are all breaking.

Solution:
- Update exception handling to look for litellm in the exception and then abort loop gracefully.
- Added 2 different types of failing operations, one from OpenAI and another from Anthropic, to prove this works.

Result:
- We now exit our loops gracefully for all LiteLLM errors.

----

This also fixes the additional warnings because we were importing litellm in agent and litellm is the library throwing all the warnings.

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comments for PR #1965

## Overview:
This pull request focuses on modifying error handling for LiteLLM exceptions throughout the codebase, particularly enhancing the handling for authentication and Anthropic-specific errors. The updates span across the following files: `agent.py`, `crew_agent_executor.py`, and `agent_test.py`.

## General Insights:
The changes made in this PR aim to streamline error handling by replacing specific exceptions with a more generic approach using `BaseLLMException`. While this brings about a more uniform handling strategy, it poses certain risks, such as potential loss of clarity regarding specific error types if not paired with adequate logging and documentation.

### File-Specific Recommendations:

#### 1. **`src/crewai/agent.py`**
- **Key Changes**:
  - Removed unused imports, which enhances code readability.
  - Transitioned from `LiteLLMAuthenticationError` to `BaseLLMException`, which may mask distinct issues related specifically to authentication without appropriate logging.

- **Improvements Suggested**:
  - **Better Exception Type Logging**:
    ```python
    logger.error(f"LiteLLM error occurred: {str(e)}")
    ```
  - **Type Hinting**: Enhance clarity in method definitions:
    ```python
    def execute_task(self, task: Task, context: Optional[str] = None) -> str:
    ```

#### 2. **`src/crewai/agents/crew_agent_executor.py`**
- **Key Changes**:
  - Refined the exception handling structure in `_invoke_loop` to provide more robust error management and prevent silent failures.
  - Removed direct handling of `OutputParserException`.

- **Improvements Suggested**:
  - **Enhanced Exception Handling Structure**:
    ```python
    try:
        # ... existing code ...
    except BaseLLMException as e:
        logger.error(f"LiteLLM error: {str(e)}")
        raise e
    ```
  - **Documentation Updates**: Ensure docstrings are up-to-date, signifying all possible exceptions that may be raised during execution:
    ```python
    """
    Raises:
        BaseLLMException: LiteLLM-related errors
    """
    ```

#### 3. **`tests/agent_test.py`**
- **Key Changes**:
  - Introduced comprehensive coverage for new error handling scenarios, verifying both authentication and Anthropic errors.
  - Employed proper use of mocks and assertions to simulate various exception conditions.

- **Improvements Suggested**:
  - **Categorize Tests**: Use markers to categorize tests effectively for easier identification:
    ```python
    @pytest.mark.error_handling
    def test_litellm_auth_error_handling():
        """Test authentication error handling."""
    ```
  - **Include More Edge Cases**: Consider various exception scenarios, such as multiple consecutive errors to ensure comprehensive testing:
    ```python
    def test_litellm_multiple_errors():
        ...
    ```

## General Recommendations:
1. **Error Logging**: Introduce extensive logging across the new error handling code to facilitate debugging and monitor patterns of errors encountered.
2. **Consistent Documentation**: Ensure all methods have updated docstrings that accurately reflect the behavior and exceptions thrown.
3. **Expanded Type Hints**: Increase type hint usage throughout the code to enhance clarity about the expected data types, benefiting maintenance and future development.
4. **Error Constants File**: Consider creating a centralized constants file for managing error messages and retry limits, promoting cleaner main code.

The overall modifications enhance the error handling structure, particularly for LiteLLM-related exceptions. They make the code more maintainable and robust in addressing diverse error scenarios, which is critical for ensuring system reliability and clarity in handling exceptions.

File: tests/agent_test.py [hunk 1]
```diff
@@ -1639,7 +1639,7 @@ def test_litellm_auth_error_handling():
 
 def test_crew_agent_executor_litellm_auth_error():
     """Test that CrewAgentExecutor handles LiteLLM authentication errors by raising them."""
-    from litellm import AuthenticationError as LiteLLMAuthenticationError
+    from litellm.exceptions import AuthenticationError
```
Comment (lorenzejay) [inline tests/agent_test.py:1642]: nice!

File: tests/agent_test.py [hunk 2]
```diff
@@ -1689,14 +1689,53 @@ def test_crew_agent_executor_litellm_auth_error():
...
-    # Verify error handling
-    mock_printer.assert_any_call(
-        content="An unknown error occurred. Please check the details below.",
-        color="red",
-    )
+    # Verify error handling messages
+    error_message = f"Error during LLM call: {str(mock_llm_call.side_effect)}"
     mock_printer.assert_any_call(
-        content="Error details: litellm.AuthenticationError: Invalid API key",
+        content=error_message,
```
Comment (lorenzejay) [inline tests/agent_test.py:1695]: lgtm

**Review by lorenzejay — APPROVED** [review]

**Review by lorenzejay — APPROVED** [review]
clean
